# Multithreading practice for web request 

In this practice, I am going to try on multithreading impact on several websites loading efficiency. Since recently I moved to Belgium for my study, I put mostly the furnitures that I am interested in for the web urls.

**The trial will be done in different situation.**


In [59]:
import urllib.request
from concurrent.futures import ThreadPoolExecutor

In [60]:
urls = [
    'https://order.ikea.com/be/nl/checkout/shoppingbag/',
    'https://www.ikea.com/be/nl/cat/bedden-bm003/',
    'https://www.ikea.com/be/nl/cat/bedbanken-logeerbedden-19037/',
    'https://www.ikea.com/be/nl/p/vagsjoen-badhanddoek-lichtrood-20439405/',
    'https://www.ikea.com/be/nl/p/grusblad-4-seizoenendekbed-20300776/',
    'https://www.ikea.com/be/nl/p/sivested-deurmat-donkergrijs-20382772/',
    'https://www.ikea.com/be/nl/p/vinnfar-badmat-donkerblauw-60447272/',
    'https://www.ikea.com/be/nl/p/eldboll-memoryfoam-polyesterkussen-20422933/',
    'https://www.ikea.com/be/nl/p/boeja-tafellamp-vernikkeld-bamboe-rotan-60152279/',

  ]

## Run without threading

In [61]:
%%time

results = []
for url in urls:
    with urllib.request.urlopen(url) as src:
        results.append(src)

CPU times: user 153 ms, sys: 12 ms, total: 165 ms
Wall time: 1.98 s


## Run with 2 threads

In [62]:
%%time

with ThreadPoolExecutor(2) as executor:
    results = executor.map(urllib.request.urlopen, urls)

CPU times: user 100 ms, sys: 11.5 ms, total: 112 ms
Wall time: 1.39 s


## Run with 4 threads

In [63]:
%%time

with ThreadPoolExecutor(4) as executor:
    results = executor.map(urllib.request.urlopen, urls)

CPU times: user 118 ms, sys: 11.3 ms, total: 129 ms
Wall time: 340 ms


## Run with 8 threads

In [64]:
%%time

with ThreadPoolExecutor(8) as executor:
    results = executor.map(urllib.request.urlopen, urls)

CPU times: user 118 ms, sys: 14.1 ms, total: 132 ms
Wall time: 243 ms


## Run with 16 threads

In [67]:
%%time

with ThreadPoolExecutor(16) as executor:
    results = executor.map(urllib.request.urlopen, urls)

CPU times: user 111 ms, sys: 10.6 ms, total: 121 ms
Wall time: 182 ms


**The wall time for these running are shown in the table below:**
    
|Thread condition | w/o multithread |  2 threads    |  4 threads    |  8 threads    |  16 threads    |
|-----------------|-----------------|---------------|---------------|---------------|----------------|
| Time spent      |  1.98 s         |   1.39 s      | 340 ms        | 243 ms        |  182 ms        |